File: src/Day5.hs
module Day5 where

import Control.Arrow ((***))
import qualified Data.Set as Set

data Bit = One | Zero

main :: IO ()
main = do
    seatIds <- map (seatId . codeToRowColumn) . lines <$> readFile "puzzle-inputs/5.txt"
    putStrLn $
        unlines
            [ "Day 5"
            , "Highest seat id " <> show (maximum seatIds)
            , "My seat number is " <> show (mySeatNumber seatIds)
            ]

seatId :: (Int , Int) -> Int
seatId (row , col) = row * 8 + col

codeToRowColumn :: String -> (Int , Int)
codeToRowColumn =
    both binaryToDecimal
        . (map rowCharToBit *** map colCharToBit)
        . span (`elem` "FB")

both f (a , b) = (f a , f b)

rowCharToBit 'F' = Zero
rowCharToBit 'B' = One

colCharToBit 'L' = Zero
colCharToBit 'R' = One

binaryToDecimal :: [Bit] -> Int
binaryToDecimal = foldr step 0 . zip [0 ..] . reverse
    where
        step (pos , One) acc = acc + 2 ^ pos
        step (_ , Zero) acc = acc

mySeatNumber :: [Int] -> Int
mySeatNumber seatList = head $ filter hasNeighbours missingSeats
    where
        seats = Set.fromList seatList
        Just minId = Set.lookupMin seats
        Just maxId = Set.lookupMax seats
        missingSeats = filter (not . (`Set.member` seats)) [minId , minId + 1 .. maxId]
        hasNeighbours sid = Set.member (sid - 1) seats && Set.member (sid + 1) seats

File: src/day5.rs
use std::collections::BTreeSet;
use std::fs;

pub fn main() {
    let file = fs::read_to_string("puzzle-inputs/5.txt").expect("Unable to read input file");
    let seat_ids: Vec<u32> = file.lines().map(seat_id).collect();
    let max_seat_id = seat_ids.iter().max().unwrap();
    println!("Max seat id: {}", max_seat_id);

    let min_id = *seat_ids.iter().min().unwrap();
    let max_id = *seat_ids.iter().max().unwrap();
    for i in min_id..max_id {
        let this = seat_ids.contains(&i);
        let prev = seat_ids.contains(&(i - 1));
        let next = seat_ids.contains(&(i + 1));
        if !this && prev && next {
            println!("My seat ID: {}", i);
        }
    }
}

fn seat_id(input: &str) -> u32 {
    let from_char = |c| match c {
        'L' => 0,
        'R' => 1,
        'F' => 0,
        'B' => 1,
        _ => panic!("Unexpected char {}", c),
    };
    let row = input[0..7].chars().fold(0, |acc, c| acc * 2 + from_char(c));
    let col = input[7..].chars().fold(0, |acc, c| acc * 2 + from_char(c));
    row * 8 + col
}
